doc/man/man1/sstat.1
.TH sstat "1" "Slurm Commands" "April 2015" "Slurm Commands"

.SH "NAME"
sstat \- Display various status information
of a running job/step.

.SH "SYNOPSIS"
.BR "sstat "
[\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Status information for running jobs invoked with Slurm.
.PP
The
.BR "sstat "
command displays job status information for your analysis.
The
.BR "sstat "
command displays information pertaining to CPU, Task, Node, Resident
Set Size (RSS) and Virtual Memory (VM).
You can tailor the output with the use of the
\f3\-\-fields=\fP
option to specify the fields to be shown.
.PP
For the root user, the
.BR "sstat "
command displays job status data for any job running on the system.
.PP
For the non\-root user, the
.BR "sstat "
output is limited to the user's jobs.

.PP
NOTE:  The
.BR "sstat "
command requires that the \f3jobacct_gather\fP plugin be installed and
operational.
.PP
NOTE:  The
.BR "sstat "
command is not supported on Cray ALPS or IBM BlueGene systems.

.TP
\f3\-a\fP\f3,\fP \f3\-\-allsteps\fP
Print all steps for the given job(s) when no step is specified.

.TP
\f3\-e\fP\f3,\fP \f3\-\-helpformat\fP
Print a list of fields that can be specified with the '\-\-format' option.

.TP
\f3\-h\fP\f3,\fP \f3\-\-help\fP
Displays a general help message.

.TP
\f3\-i\fP\f3,\fP \f3\-\-pidformat\fP
Predefined format to list the pids running for each job step.
(JobId,Nodes,Pids)

.TP
\f3\-j\fP\f3,\fP \f3\-\-jobs\fP
Format is <job(.step)>. Stat this job step or comma-separated list of
job steps. This option is required.  The step portion will default to
lowest step running if not specified, unless the \-\-allsteps flag is
set where not specifying a step will result in all running steps to be
displayed.
NOTE: A step id of 'batch' will display the information about the batch step.
NOTE: A step id of 'extern' will display the information about the
extern step.  This step is only available when using PrologFlags=contain

.TP
\f3\-n\fP\f3,\fP \f3\-\-noheader\fP
No heading will be added to the output. The default action is to
display a header.

.TP
\f3\-\-noconvert\fP
Don't convert units from their original type (e.g. 2048M won't be converted to
2G).

.TP
\f3\-o\fP\f3,\fP \f3\-\-format\fP,\fP \f3\-\-fields\fP
Comma separated list of fields.
(use '\-\-helpformat' for a list of available fields).

NOTE: When using the format option for listing various fields you can put a
%NUMBER afterwards to specify how many characters should be printed.

i.e. format=name%30 will print 30 characters of field name right
justified.  A \-30 will print 30 characters left justified.

.TP
\f3\-p\fP\f3,\fP \f3\-\-parsable\fP
output will be '|' delimited with a '|' at the end

.TP
\f3\-P\fP\f3,\fP \f3\-\-parsable2\fP
output will be '|' delimited without a '|' at the end

.TP
\f3\-\-usage\fP
Display a command usage summary.

.TP
\f3\-v\fP\f3,\fP \f3\-\-verbose\fP
Primarily for debugging purposes, report the state of various
variables during processing.

.TP
\f3\-V\fP\f3,\fP \f3\-\-version\fP
Print version.


.SS "Job Status Fields"
The following are the field options:
.RS
.TP
\f3AveCPU\fP
Average (system + user) CPU time of all tasks in job.

.TP
\f3AveCPUFreq\fP
Average weighted CPU frequency of all tasks in job, in kHz.

.TP
\f3AveDiskRead\fP
Average number of bytes read by all tasks in job.

.TP
\f3AveDiskWrite\fP
Average number of bytes written by all tasks in job.

.TP
\f3AvePages\fP
Average number of page faults of all tasks in job.

.TP
\f3AveRSS\fP
Average resident set size of all tasks in job.

.TP
\f3AveVMSize\fP
Average Virtual Memory size of all tasks in job.

.TP
\f3ConsumedEnergy\fP
Total energy consumed by all tasks in job, in joules.
Note: Only in case of exclusive job allocation this value 
reflects the jobs' real energy consumption. 

.TP
\f3JobID\fP
The number of the job or job step.
It is in the form:
\f2job.jobstep\fP\c
\&.

.TP
\f3MaxDiskRead\fP
Maximum number of bytes read by all tasks in job.

.TP
\f3MaxDiskReadNode\fP
The node on which the maxdiskread occurred.

.TP
\f3MaxDiskReadTask\fP
The task ID where the maxdiskread occurred.

.TP
\f3MaxDiskWrite\fP
Maximum number of bytes written by all tasks in job.

.TP
\f3MaxDiskWriteNode\fP
The node on which the maxdiskwrite occurred.

.TP
\f3MaxDiskWriteTask\fP
The task ID where the maxdiskwrite occurred.

.TP
\f3MaxPages\fP
Maximum number of page faults of all tasks in job.

.TP
\f3MaxPagesNode\fP
The node on which the maxpages occurred.

.TP
\f3MaxPagesTask\fP
The task ID where the maxpages occurred.

.TP
\f3MaxRSS\fP
Maximum resident set size of all tasks in job.

.TP
\f3MaxRSSNode\fP
The node on which the maxrss occurred.

.TP
\f3MaxRSSTask\fP
The task ID where the maxrss occurred.

.TP
\f3MaxVMSize\fP
Maximum Virtual Memory size of all tasks in job.

.TP
\f3MaxVMSizeNode\fP
The node on which the maxvsize occurred.

.TP
\f3MaxVMSizeTask\fP
The task ID where the maxvsize occurred.

.TP
\f3MinCPU\fP
Minimum (system + user) CPU time of all tasks in job.

.TP
\f3MinCPUNode\fP
The node on which the mincpu occurred.

.TP
\f3MinCPUTask\fP
The task ID where the mincpu occurred.

.TP
\f3NTasks\fP
Total number of tasks in a job or step.

.TP
\f3PackJobID\fP
Jobid of srun first step of the jobpack.

.TP
\f3PackStepID\fP
Stepid of jobpack member.

.TP
\f3ReqCPUFreq\fP
Requested CPU frequency for the step, in kHz.

.SH "ENVIRONMENT VARIABLES"
.PP
Some \fBsstat\fR options may be set via environment variables. These
environment variables, along with their corresponding options, are listed below.
(Note: commandline options will always override these settings)
.TP 20
\fBSLURM_CONF\fR
The location of the Slurm configuration file.

.SH "EXAMPLES"

.TP
\f3sstat \-\-format=AveCPU,AvePages,AveRSS,AveVMSize,JobID \-j 11\fP
25:02.000  0K         1.37M      5.93M      9.0

.TP
\f3sstat \-p \-\-format=AveCPU,AvePages,AveRSS,AveVMSize,JobID \-j 11\fP
25:02.000|0K|1.37M|5.93M|9.0|

.SH "COPYING"
Copyright (C) 2009 Lawrence Livermore National Security.
Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.br
Copyright (C) 2010\-2013 SchedMD LLC.
.LP
This file is part of Slurm, a resource management program.
For details, see <https://slurm.schedmd.com/>.
.LP
Slurm is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.
.LP
Slurm is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

.SH "SEE ALSO"
\fBsacct\fR(1)


RELEASE_NOTES
RELEASE NOTES FOR SLURM VERSION 17.02
23 February 2017

IMPORTANT NOTES:
THE MAXJOBID IS NOW 67,108,863. ANY PRE-EXISTING JOBS WILL CONTINUE TO RUN BUT
NEW JOB IDS WILL BE WITHIN THE NEW MAXJOBID RANGE. Adjust your configured
MaxJobID value as needed to eliminate any confusion.

If using the slurmdbd (Slurm DataBase Daemon) you must update this first.
The 17.02 slurmdbd will work with Slurm daemons of version 15.08 and above.
You will not need to update all clusters at the same time, but it is very
important to update slurmdbd first and having it running before updating
any other clusters making use of it.  No real harm will come from updating
your systems before the slurmdbd, but they will not talk to each other
until you do.  Also at least the first time running the slurmdbd you need to
make sure your my.cnf file has innodb_buffer_pool_size equal to at least 64M.
You can accomplish this by adding the line

innodb_buffer_pool_size=64M

under the [mysqld] reference in the my.cnf file and restarting the mysqld. The
buffer pool size must be smaller than the size of the MySQL tmpdir. This is
needed when converting large tables over to the new database schema.

Slurm can be upgraded from version 15.08 or 16.05 to version 17.02 without loss
of jobs or other state information. Upgrading directly from an earlier version
of Slurm will result in loss of state information.

If using SPANK plugins that use the Slurm APIs, they should be recompiled when
upgrading Slurm to a new major release.

NOTE: systemd services files are installed automatically, but not enabled.
      You will need to manually enable them on the appropriate systems:
      - Controller: systemctl enable slurmctld
      - Database: systemctl enable slurmdbd
      - Compute Nodes: systemctl enable slurmd

NOTE: If you are not using Munge, but are using the "service" scripts to
      start Slurm daemons, then you will need to remove this check from the
      etc/slurm*service scripts.

NOTE: If you are upgrading with any jobs from 14.03 or earlier
      (i.e. quick upgrade from 14.03 -> 15.08 -> 17.02) you will need
      to wait until after those jobs are gone before you upgrade to 17.02.

NOTE: If you interact with any memory values in a job_submit plugin, you will
      need to test against NO_VAL64 instead of NO_VAL, and change your printf
      format as well.

HIGHLIGHTS
==========
 -- Added infrastructure for managing workload across a federation of clusters.
    (partial functionality in version 17.02, fully operational in May 2017)
 -- In order to support federated jobs, the MaxJobID configuration parameter
    default value has been reduced from 2,147,418,112 to 67,043,328 and its
    maximum value is now 67,108,863. Upon upgrading, any pre-existing jobs that
    have a job ID above the new range will continue to run and new jobs will get
    job IDs in the new range.
 -- Added "MailDomain" configuration parameter to qualify email addresses.
 -- Automatically clean up task/cgroup cpuset and devices cgroups after steps
    are completed.
 -- Added burst buffer support for job arrays. Added new SchedulerParameters
    configuration parameter of bb_array_stage_cnt=# to indicate how many pending
    tasks of a job array should be made available for burst buffer resource
    allocation.
 -- Added new sacctmgr commands: "shutdown" (shutdown the server), "list stats"
    (get server statistics) "clear stats" (clear server statistics).
 -- The database index for jobs is now 64 bits.  If you happen to be close to
    4 billion jobs in your database you will want to update your slurmctld at
    the same time as your slurmdbd to prevent roll over of this variable as
    it is 32 bit previous versions of Slurm.
 -- All memory values (in MB) are now 64 bit. Previously, nodes with more than
    2 TB of memory would not schedule or enforce memory limits correctly.
 -- Removed AIX, BlueGene/L and BlueGene/P support.
 -- Removed sched/wiki and sched/wiki2 plugins and associated code.
 -- Added PrologFlags=Serial to disable concurrent execution of prolog/epilog
    scripts.

RPMBUILD CHANGES
================
 -- Removed separate slurm_blcr package. If Slurm is built with BLCR support,
    the files will now be part of the main Slurm packages.
 -- Replace sjstat, seff and sjobexit RPM packages with a single "contribs"
    package.
 -- Prevent installation of both init scripts and systemd service files.
    Note that systemd services are not automatically enabled at this time,
    unlike the init scripts which were automatically enabled with chkconfig.

CONFIGURATION FILE CHANGES (see man appropriate man page for details)
=====================================================================
 -- Added new "TRESWeights" configuration parameter to "NodeName" line for
    calculating how busy a node is. Currently only used for federation
    configurations.
 -- Added new "PriorityFlags" option of "INCR_ONLY", which prevents a job's
    priority from being decremented.
 -- Added "SchedulerParameters" configuration parameter of "default_gbytes",
    which treats numeric only (no suffix) value for memory and tmp disk space as
    being in units of Gigabytes. Mostly for compatability with LSF.
 -- Added "BootTime" configuration parameter to knl.conf file to optimize
    resource allocations with respect to required node reboots.
 -- Added "SchedulingParameters" option of "bf_job_part_count_reserve". Jobs
    below the specified threshold will not have resources reserved for them.
 -- Added "SchedulerParameters" option of "spec_cores_first" to select
    specialized cores from the lowest rather than highest number cores and
    sockets.
 -- Added "PrologFlags" option of "Serial" to disable concurrent launch of
    Prolog and Epilog scripts.
 -- Removed "SchedulerRootFilter" configuration parameter.
 -- Added "SyscfgTimeout" parameter to knl.conf configuration file.

COMMAND CHANGES (see man pages for details)

src/sstat/sstat.c
/*****************************************************************************\
 *  sstat.c - job accounting reports for SLURM's slurmdb/log plugin
 *****************************************************************************
 *  Copyright (C) 2008 Lawrence Livermore National Security.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Morris Jette <jette1@llnl.gov>
 *  CODE-OCEC-09-009. All rights reserved.
 *
 *  This file is part of SLURM, a resource management program.
 *  For details, see <https://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  SLURM is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  SLURM is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with SLURM; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#include "config.h"

#include "sstat.h"

int _do_stat(uint32_t jobid, uint32_t stepid, char *nodelist,
	     uint32_t req_cpufreq_min, uint32_t req_cpufreq_max,
	     uint32_t req_cpufreq_gov, uint16_t use_protocol_ver, 
	     uint32_t packjobid, uint32_t packstepid);

/*
 * Globals
 */
sstat_parameters_t params;
print_field_t fields[] = {
	{10, "AveCPU", print_fields_str, PRINT_AVECPU},
	{10, "AveCPUFreq", print_fields_str, PRINT_ACT_CPUFREQ},
	{12, "AveDiskRead", print_fields_str, PRINT_AVEDISKREAD},
	{12, "AveDiskWrite", print_fields_str, PRINT_AVEDISKWRITE},
	{10, "AvePages", print_fields_str, PRINT_AVEPAGES},
	{10, "AveRSS", print_fields_str, PRINT_AVERSS},
	{10, "AveVMSize", print_fields_str, PRINT_AVEVSIZE},
	{14, "ConsumedEnergy", print_fields_str, PRINT_CONSUMED_ENERGY},
	{17, "ConsumedEnergyRaw", print_fields_double,
	 PRINT_CONSUMED_ENERGY_RAW},
	{-12, "JobID", print_fields_str, PRINT_JOBID},
	{12, "MaxDiskRead", print_fields_str, PRINT_MAXDISKREAD},
	{15, "MaxDiskReadNode", print_fields_str, PRINT_MAXDISKREADNODE},
	{15, "MaxDiskReadTask", print_fields_uint, PRINT_MAXDISKREADTASK},
	{12, "MaxDiskWrite", print_fields_str, PRINT_MAXDISKWRITE},
	{16, "MaxDiskWriteNode", print_fields_str, PRINT_MAXDISKWRITENODE},
	{16, "MaxDiskWriteTask", print_fields_uint, PRINT_MAXDISKWRITETASK},
	{8, "MaxPages", print_fields_str, PRINT_MAXPAGES},
	{12, "MaxPagesNode", print_fields_str, PRINT_MAXPAGESNODE},
	{14, "MaxPagesTask", print_fields_uint, PRINT_MAXPAGESTASK},
	{10, "MaxRSS", print_fields_str, PRINT_MAXRSS},
	{10, "MaxRSSNode", print_fields_str, PRINT_MAXRSSNODE},
	{10, "MaxRSSTask", print_fields_uint, PRINT_MAXRSSTASK},
	{10, "MaxVMSize", print_fields_str, PRINT_MAXVSIZE},
	{14, "MaxVMSizeNode", print_fields_str, PRINT_MAXVSIZENODE},
	{14, "MaxVMSizeTask", print_fields_uint, PRINT_MAXVSIZETASK},
	{10, "MinCPU", print_fields_str, PRINT_MINCPU},
	{10, "MinCPUNode", print_fields_str, PRINT_MINCPUNODE},
	{10, "MinCPUTask", print_fields_uint, PRINT_MINCPUTASK},
	{20, "Nodelist", print_fields_str, PRINT_NODELIST},
	{8, "NTasks", print_fields_uint, PRINT_NTASKS},
	{20, "PackJobID", print_fields_uint, PRINT_PACKJOBID},
	{20, "PackStepID", print_fields_uint, PRINT_PACKSTEPID},
	{20, "Pids", print_fields_str, PRINT_PIDS},
	{10, "ReqCPUFreq", print_fields_str, PRINT_REQ_CPUFREQ_MIN}, /*vestigial*/
	{13, "ReqCPUFreqMin", print_fields_str, PRINT_REQ_CPUFREQ_MIN},
	{13, "ReqCPUFreqMax", print_fields_str, PRINT_REQ_CPUFREQ_MAX},
	{13, "ReqCPUFreqGov", print_fields_str, PRINT_REQ_CPUFREQ_GOV},
	{0, NULL, NULL, 0}};

List jobs = NULL;
slurmdb_job_rec_t job;
slurmdb_step_rec_t step;
List print_fields_list = NULL;
ListIterator print_fields_itr = NULL;
int field_count = 0;

int _do_stat(uint32_t jobid, uint32_t stepid, char *nodelist,
	     uint32_t req_cpufreq_min, uint32_t req_cpufreq_max,
	     uint32_t req_cpufreq_gov, uint16_t use_protocol_ver, 
	     uint32_t packjobid, uint32_t packstepid)

{
	job_step_stat_response_msg_t *step_stat_response = NULL;
	int rc = SLURM_SUCCESS;
	ListIterator itr;
	slurmdb_stats_t temp_stats;
	job_step_stat_t *step_stat = NULL;
	int ntasks = 0;
	int tot_tasks = 0;
	hostlist_t hl = NULL;

	debug("requesting info for job %u.%u", jobid, stepid);
	if ((rc = slurm_job_step_stat(jobid, stepid, nodelist, use_protocol_ver,
				      &step_stat_response)) != SLURM_SUCCESS) {
		if (rc == ESLURM_INVALID_JOB_ID) {
			debug("job step %u.%u has already completed",
			      jobid, stepid);
		} else {
			error("problem getting step_layout for %u.%u: %s",
			      jobid, stepid, slurm_strerror(rc));
		}
		slurm_job_step_pids_response_msg_free(step_stat_response);
		return rc;
	}

	memset(&job, 0, sizeof(slurmdb_job_rec_t));
	job.jobid = jobid;

	memset(&step, 0, sizeof(slurmdb_step_rec_t));

	memset(&temp_stats, 0, sizeof(slurmdb_stats_t));
	temp_stats.cpu_min = NO_VAL;
	memset(&step.stats, 0, sizeof(slurmdb_stats_t));
	step.stats.cpu_min = NO_VAL;

	step.job_ptr = &job;
	step.stepid = stepid;
	step.nodes = xmalloc(BUF_SIZE);
	step.req_cpufreq_min = req_cpufreq_min;
	step.req_cpufreq_max = req_cpufreq_max;
	step.req_cpufreq_gov = req_cpufreq_gov;
	step.stepname = NULL;
	step.state = JOB_RUNNING;
	step.packstepid[0] = packjobid;
	step.packstepid[1] = packstepid;

	hl = hostlist_create(NULL);
	itr = list_iterator_create(step_stat_response->stats_list);
	while ((step_stat = list_next(itr))) {
		if (!step_stat->step_pids || !step_stat->step_pids->node_name)
			continue;
		if (step_stat->step_pids->pid_cnt > 0 ) {
			int i;
			for(i=0; i<step_stat->step_pids->pid_cnt; i++) {
				if (step.pid_str)
					xstrcat(step.pid_str, ",");
				xstrfmtcat(step.pid_str, "%u",
					   step_stat->step_pids->pid[i]);
			}
		}

		if (params.pid_format) {
			step.nodes = step_stat->step_pids->node_name;
			print_fields(&step);
			xfree(step.pid_str);
		} else {
			hostlist_push_host(hl, step_stat->step_pids->node_name);
			ntasks += step_stat->num_tasks;
			if (step_stat->jobacct) {
				jobacctinfo_2_stats(&temp_stats,
						    step_stat->jobacct);
				aggregate_stats(&step.stats, &temp_stats);
			}
		}
	}
	list_iterator_destroy(itr);
	slurm_job_step_pids_response_msg_free(step_stat_response);
	/* we printed it out already */
	if (params.pid_format)
		return rc;

	hostlist_sort(hl);
	hostlist_ranged_string(hl, BUF_SIZE, step.nodes);
	hostlist_destroy(hl);
	tot_tasks += ntasks;

	if (tot_tasks) {
		step.stats.cpu_ave /= (double)tot_tasks;
		step.stats.rss_ave /= (double)tot_tasks;
		step.stats.vsize_ave /= (double)tot_tasks;
		step.stats.pages_ave /= (double)tot_tasks;
		step.stats.disk_read_ave /= (double)tot_tasks;
		step.stats.disk_write_ave /= (double)tot_tasks;
		step.stats.act_cpufreq /= (double)tot_tasks;
		step.ntasks = tot_tasks;
	}

	print_fields(&step);

	return rc;
}

int main(int argc, char **argv)
{
	ListIterator itr = NULL;
	uint32_t req_cpufreq_min = NO_VAL;
	uint32_t req_cpufreq_max = NO_VAL;
	uint32_t req_cpufreq_gov = NO_VAL;
	uint32_t stepid = NO_VAL;
	uint32_t packjobid = NO_VAL;
	uint32_t packstepid = NO_VAL;
	slurmdb_selected_step_t *selected_step = NULL;

#ifdef HAVE_ALPS_CRAY
	error("The sstat command is not supported on Cray systems");
	return 1;
#endif
#ifdef HAVE_BG
	error("The sstat command is not supported on IBM BlueGene systems");
	return 1;
#endif

	slurm_conf_init(NULL);
	print_fields_list = list_create(NULL);
	print_fields_itr = list_iterator_create(print_fields_list);

	parse_command_line(argc, argv);
	if (!params.opt_job_list || !list_count(params.opt_job_list)) {
		error("You didn't give me any jobs to stat.");
		return 1;
	}

	print_fields_header(print_fields_list);
	itr = list_iterator_create(params.opt_job_list);
	while ((selected_step = list_next(itr))) {
		char *nodelist = NULL;
		bool free_nodelist = false;
		uint16_t use_protocol_ver = (uint16_t)NO_VAL;
		if (selected_step->stepid == SSTAT_BATCH_STEP) {
			/* get the batch step info */
			job_info_msg_t *job_ptr = NULL;
			hostlist_t hl;

			if (slurm_load_job(
				    &job_ptr, selected_step->jobid, SHOW_ALL)) {
				error("couldn't get info for job %u",
				      selected_step->jobid);
				continue;
			}

			use_protocol_ver =
				job_ptr->job_array[0].start_protocol_ver;
			stepid = SLURM_BATCH_SCRIPT;
			hl = hostlist_create(job_ptr->job_array[0].nodes);
			nodelist = hostlist_shift(hl);
			free_nodelist = true;
			hostlist_destroy(hl);
			slurm_free_job_info_msg(job_ptr);
		} else if (selected_step->stepid == SSTAT_EXTERN_STEP) {
			/* get the extern step info */
			job_info_msg_t *job_ptr = NULL;

			if (slurm_load_job(
				    &job_ptr, selected_step->jobid, SHOW_ALL)) {
				error("couldn't get info for job %u",
				      selected_step->jobid);
				continue;
			}
			use_protocol_ver =
				job_ptr->job_array[0].start_protocol_ver;
			stepid = SLURM_EXTERN_CONT;
			nodelist = job_ptr->job_array[0].nodes;
			slurm_free_job_info_msg(job_ptr);
		} else if (selected_step->stepid != NO_VAL) {
			stepid = selected_step->stepid;
		} else if (params.opt_all_steps) {
			job_step_info_response_msg_t *step_ptr = NULL;
			int i = 0;
			if (slurm_get_job_steps(
				    0, selected_step->jobid, NO_VAL,
				    &step_ptr, SHOW_ALL)) {
				error("couldn't get steps for job %u",
				      selected_step->jobid);
				continue;
			}

			for (i = 0; i < step_ptr->job_step_count; i++) {
				_do_stat(selected_step->jobid,
					 step_ptr->job_steps[i].step_id,
					 step_ptr->job_steps[i].nodes,
					 step_ptr->job_steps[i].cpu_freq_min,
					 step_ptr->job_steps[i].cpu_freq_max,
					 step_ptr->job_steps[i].cpu_freq_gov,
					 step_ptr->job_steps[i].
					 start_protocol_ver);
					 step_ptr->job_steps[i].packstepid[0],
					 step_ptr->job_steps[i].packstepid[1]);
			}
			slurm_free_job_step_info_response_msg(step_ptr);
			continue;
		} else {
			/* get the first running step to query against. */
			job_step_info_response_msg_t *step_ptr = NULL;
			job_step_info_t *step_info;

			if (slurm_get_job_steps(
				    0, selected_step->jobid, NO_VAL,
				    &step_ptr, SHOW_ALL)) {
				error("couldn't get steps for job %u",
				      selected_step->jobid);
				continue;
			}
			if (!step_ptr->job_step_count) {
				error("no steps running for job %u",
				      selected_step->jobid);
				continue;
			}

			/* If the first step is the extern step lets
			 * just skip it.  They should ask for it
			 * directly.
			 */
			if ((step_ptr->job_steps[0].step_id ==
			    SLURM_EXTERN_CONT) && step_ptr->job_step_count > 1)
				step_info = ++step_ptr->job_steps;
			else
				step_info = step_ptr->job_steps;
			stepid = step_info->step_id;
			nodelist = step_info->nodes;
			req_cpufreq_min = step_info->cpu_freq_min;
			req_cpufreq_max = step_info->cpu_freq_max;
			req_cpufreq_gov = step_info->cpu_freq_gov;
			use_protocol_ver = step_info->start_protocol_ver;
			packjobid = step_info->job_steps[0].packstepid[0];
			packstepid = step_info->job_steps[0].packstepid[1];
		}

		_do_stat(selected_step->jobid, stepid, nodelist,
			 req_cpufreq_min, req_cpufreq_max, req_cpufreq_gov,
			 use_protocol_ver, packjobid, packstepid);
		if (free_nodelist && nodelist)
			free(nodelist);
	}
	list_iterator_destroy(itr);

	xfree(params.opt_field_list);
	FREE_NULL_LIST(params.opt_job_list);
	if (print_fields_itr)
		list_iterator_destroy(print_fields_itr);
	FREE_NULL_LIST(print_fields_list);

	return 0;
}


